### PR TITLE
Adding annotation capabilities to SignedRoot macro

### PR DESCRIPTION
**Pull request by seansu4you87**
## Issue Addressed

https://github.com/sigp/lighthouse/issues/287

## Proposed Changes

Lets the user annotate fields of a struct to skip for signed root
hashing.

### CODE REVIEW

**Comment by seansu4you87** [comment]
@mjkeating @paulhauner Can you PTAL?

**Comment by mjkeating** [comment]
I'm wondering, do we really need the separate `SignedRoot` macro with its helper functions, etc? 

Since the signature is just a `hash_tree_root` of the container without the the final 'signature' field, it seems that if we just add this filtering functionality (removing the signature field) to `get_tree_hashable_named_field_idents`, the HashTree macro would work fine, nor would it break current TreeHash functionality.

**Comment by seansu4you87** [comment]
@mjkeating I agree, the code is very similar.  It seems like we can remove the `SignedRoot` trait completely, or if we want to keep that trait, we can at least derive it the same way we derive `TreeHash`.

```rust
#[derive(SignedRoot)] // automatically derives TreeHash (we can do that with macros right?)
struct Casper {
  friendly: bool,
  #[tree_hash(skip)] // alternatively, #[signed_root(skip)] would also work
  friends: Vec<u8>
  dead: bool
}
```

The only downside, is that we can't have a `hash_tree_root` function AND a `signed_root` function that return different outputs, without doing something like what we have now (basically two orthogonal sets of hashed fields).

So to summarize the options, we have:
1. Kill `SignedRoot` trait completely, only use `TreeHash`
2. Keep `SignedRoot` trait, but derive it with `TreeHash`, no more orthogonal hashing
3. Keep what is presented now, with two separate implementations of struct hashing

What do you think?  cc @paulhauner

**Comment by paulhauner** [comment]
> The only downside, is that we can't have a `hash_tree_root` function AND a `signed_root` function that return different outputs, without doing something like what we have now (basically two orthogonal sets of hashed fields).

Yep, I agree. The options seem to be:

1. Keep `SignedRoot`.
2. Add `signed_root()` to `TreeHash`.

I'm leaning towards (1) as it keeps `TreeHash` clean. There's some code duplication here, however it's quite minimal and only exists in the derive macro. Another advantage for (1) is that it is more explicit -- we must declare when an object should be `SignedRoot`-capable. This makes it ever so slightly more difficult to accidentally use a `SignedRoot` when you should be using a `TreeHash`.

I'm in the thick of getting the single-client test-net up so I won't be able to go through this in detail until Monday, however after a quick glance it looks well implemented with clean and tidy tests. Thank you!

@seansu4you87 I'm not sure if you're already across this but we're looking to [change licenses](https://github.com/sigp/lighthouse/issues/326). Please let us know if you would have issues if this code were merged and re-licensed under Apache 2.

**Comment by seansu4you87** [comment]
@paulhauner Great, if the desire is to have these separate concepts of a tree hash vs a signed root, then I agree, (1) seems like the best option.  I will look for all usage of `SignedRoot`, and change it to use an explicit annotation on the signature fields to skip, instead of implicit field names

btw, I have just agreed to the re-licensing

**Comment by seansu4you87** [comment]
@paulhauner Ok ready for a more thorough review when you have a chance.  I changed the tests so that it more explicitly tests for the orthogonal set membership for tree vs signed hashing, which caught a bug.  Also added `#[signed_root(skip_hashing)]` annotations everywhere

**Comment by paulhauner** [comment]
This is great, I have nothing to add. Thank you!

Apologies again for the delayed response, the past couple of weeks have been quite disrupted for us due to a mixture of a sprint for testnet, taking leave and dealing with EDCON Sydney.

We have a conflict with `master`. It has changed significantly over the past week so I would suggest merge/rebase with it and ensure the tests pass. Hopefully Jenkins will pass too, we've been having some difficulties with it.

Thanks again!

**Comment by protolambda** [comment]
Was lurking :eyes: in the lighthouse gitter and thought I could add something here: Adding tags/annotations to your structs to define what is ignored and what is not for signed-root works great, example in ZRNT can be found here: https://github.com/protolambda/zrnt/blob/c8e0a5deeba144749c451ba5251e3db513b0e0b7/eth2/beacon/block.go#L15

**Comment by paulhauner** [comment]
Hey @seansu4you87, the build is failing on rust-fmt.

It should pass if you run `cargo fmt --all` and commit the changes :)

**Comment by seansu4you87** [comment]
@paulhauner Got the PR all green =]

Lemme know if there's anything else you want, but as is, it should be good to go

**Comment by paulhauner** [comment]
Thank you! Excited to have a new contributor!
